Add Ubuntu Desktop workflows (#229)
ubuntu/ubuntu-project-docs: docs/contributors/setup/set-up-git-for-ubuntu-desktop-work.md

---
relatedlinks: "[PackagingWithGit &#32; - &#32; Debian Wiki](https://wiki.debian.org/PackagingWithGit)"
---

(set-up-git-for-ubuntu-desktop-work)=
# Set up Git for Ubuntu Desktop work

> The [Desktop Team](https://library.canonical.com/our-organisation/ubuntu-engineering/desktop) uses Git and Git Build Package (`gbp`) to maintain GNOME packages for Ubuntu. Let's set up your system so that you can contribute to GNOME on Ubuntu Desktop.

Reviewer: The Library is internal to Canonical. These docs are public.
```suggestion
The [Desktop Team](https://discourse.ubuntu.com/c/project/desktop/8) uses Git and Git Build Package (`gbp`) to maintain GNOME packages for Ubuntu. Let's set up your system so that you can contribute to GNOME on Ubuntu Desktop.
```

> We'll use the Settings application as an example. Internally, the application is known as the GNOME Control Center.

Reviewer: ```suggestion
We use the Settings application as an example. Internally, the application is known as the GNOME Control Center.
```

## Initial setup

1. Install the basic packages for working in Git and Ubuntu packaging:

    ```{terminal}
    :copy:
    :host:
    :dir:
    :user:
    :input: sudo apt install git git-buildpackage ubuntu-dev-tools gnupg
    ```

>     For details about the packaging environment, see [Getting Set Up](https://canonical-ubuntu-packaging-guide.readthedocs-hosted.com/en/1.0/tutorial/getting-set-up.html) in the Ubuntu Packaging Guide.

Reviewer: The Packaging Guide is deprecated. Use resources from within these docs, please.
```suggestion
    For details about the packaging environment, see {ref}`how-to-set-up-for-ubuntu-development`.
```

    :dir:
    :user:
    :input: sudo apt install gnupg pbuilder ubuntu-dev-tools apt-file
    ```

    For details about the packaging environment, see [Getting Set Up](https://canonical-ubuntu-packaging-guide.readthedocs-hosted.com/en/1.0/tutorial/getting-set-up.html) in the Ubuntu Packaging Guide.

3. In the Software & Updates application, enable {guilabel}`Source code` on the {guilabel}`Ubuntu Software` tab.

    This provides information about the source repository of every package so that you can easily clone it.

4. Enable the `salsa:name-or-team/repo` and `salsa-gnome:repo` short format for Salsa Git repositories.

    Add the following configuration to your `~/.config/git/config` file:

    ```{code-block} ini
    :caption: `~/.config/git/config`

    [url "git@salsa.debian.org:"]
        insteadof = salsa:

    [url "git@salsa.debian.org:gnome-team/"]
        insteadof = salsa-gnome:
    ```

5. Automatically sign tags with your GPG key.

    Add the following configuration to your `~/.gbp.conf` file:

    ```{code-block} ini
    :caption: `~/.gbp.conf`

    [buildpackage]
    sign-tags = True
    keyid = 0xGPGKEYID
    ```

>     Replace `GPGKEYID` with your GPG key ID, which is the last eight digits of your public GPG key.

Author: GPG signing was in the optional section until our review meeting where I think either Alessandro or Nathan told me to make it mandatory. I guess there are some projects that require it, and it's better to enable it by default than to find out later that you need to rewrite your Git history somewhere. 

Adding the GPG command and clarifying the short key.

    :host:
    :dir:
    :user:
    :input: debcheckout --git-track=* gnome-control-center

    declared git repository at https://salsa.debian.org/gnome-team/gnome-control-center/
    git clone https://salsa.debian.org/gnome-team/gnome-control-center -b ubuntu/latest gnome-control-center ...
    […]
    ```

    If the `Vcs-Git` field was missing or incorrect, you could clone the repository manually using `gbp`:

    ```{terminal}
    :copy:
    :host:
    :dir:
    :user:
    :input: gbp clone salsa-gnome:gnome-control-center

    gbp:info: Cloning from 'salsa-gnome:gnome-control-center'
    ```

## Configure branches

1. Switch to the cloned repository:

    ```{terminal}
    :copy:
    :host:
    :dir:
    :user:
    :input: cd gnome-control-center
    ```

1. Enable Git to push tags automatically because `gbp` relies on them.

    Apply this as local configuration in the cloned repository:

    ```{terminal}
    :copy:
    :host:
    :dir: gnome-control-center
    :user:
    :input: git config push.followTags true
    ```

1. Add the upstream GNOME remote repository and call it `upstreamvcs`:

    ```{terminal}
    :copy:
    :host:
    :dir: gnome-control-center
    :user:

>     :input: git remote add -f upstreamvcs git@ssh.gitlab.gnome.org:GNOME/gnome-control-center.git

Reviewer 2: Note that this URI requires being registered in gitlab GNOME, maybe you can warn or just suggest to use the https URI: https://gitlab.gnome.org/GNOME/gnome-control-center.git

    :::

1. Check that you now have the following three branches:

    ```{terminal}
    :copy:
    :host:
    :dir: gnome-control-center
    :user:
    :input: git branch -vv

      pristine-tar             a96c0e2e98 [origin/pristine-tar] pristine-tar data for gnome-control-center_49.0.orig.tar.xz
    * ubuntu/latest            a8640ab8a4 [origin/ubuntu/latest] debian/salsa-ci: Enable for ubuntu

>       upstream/latest          4db8b3a502 [origin/upstream/latest] New upstream version 49.0

Reviewer: ```suggestion
      pristine-tar      a96c0e2e98 [origin/pristine-tar] pristine-tar data for gnome-control-center_49.0.orig.tar.xz
    * ubuntu/latest     a8640ab8a4 [origin/ubuntu/latest] debian/salsa-ci: Enable for ubuntu
      upstream/latest   4db8b3a502 [origin/upstream/latest] New upstream version 49.0
```

    ```

1. Check that you have the following two remotes:

    ```{terminal}
    :copy:
    :host:
    :dir: gnome-control-center
    :user:
    :input: git remote

    origin
    upstreamvcs
    ```

## Examine the branches and repositories

We use a layout with two remote repositories. Let's take a look.

Run these commands in the `gnome-control-center` repository that we cloned earlier.

```{terminal}
:copy:
:host:
:dir: gnome-control-center
:user:

> :input: git remote show

Reviewer 2: ```suggestion
:input: git remote show -v
```

> :input: git remote show -v

Reviewer: The `show` subcommand doesn't have the `-v` option. Did you mean:
```suggestion
:input: git remote -v
```

```

The command lists the following remote repositories:

`origin`
: Points to the Salsa repository. The default for pull and push without any argument.

`upstreamvcs`
: Points to the upstream GNOME repository, to easily cherry-pick upstream fixes.

    This remote is called `upstreamvcs` to not be confused with local branch names, which might be called `debian/<branch-name>` or `upstream/<branch-name>`. Instead of having a local `debian/latest` branch tracking the `debian/debian/latest` remote, we have the `debian/latest` local branch tracking the `origin/debian/latest` remote, which is easier to understand.

:::{note}
The `debian/latest` and `ubuntu/latest` branches were previously called `debian/master` and `ubuntu/master`, respectively. We renamed them on 4 September 2023 to use more inclusive naming and more closely follow [DEP-14](https://dep-team.pages.debian.net/deps/dep14/).
:::

These names of remote repositories are local to your system. You can rename them without affecting the content of the remote repositories.

> ### Salsa remote branches

Reviewer: ```suggestion

### Salsa remote branches
```

Let's look at the various branches in the Salsa remote repository:

```{terminal}
:copy:
:host:
:dir: gnome-control-center
:user:
:input: git remote show origin

* remote origin
  […]
  HEAD branch: debian/latest
  Remote branches:
    pristine-tar             tracked
    ubuntu/latest            tracked
    upstream/latest          tracked
  […]
```

`ubuntu/latest`

> : This branch is the content of the latest Ubuntu development release. Work ready to be uploaded or uploaded in a development release mostly happens here. It's the default branch.

Author: Added.

:::{admonition} TODO
:class: attention
That's a long sentence. Who fetches the content? Who injects the history? What does it mean to inject history in Git?
:::

Let's browse the content of the `upstreamvcs` repository:

```{terminal}
:copy:
:host:
:dir: gnome-control-center
:user:
:input: git remote show upstreamvcs

* remote upstreamvcs
[…]
  HEAD branch: main
  Remote branches:
[…]
    gnome-49                                  tracked
[…]
```

You can check out any of those branch locally. We have the local `upstream/latest` branch tracking the `main` branch of the `upstreamvcs` repository. We can track any additional branches as needed for cherry-picking fixes.

:::{admonition} TODO
:class: attention
That's not true though. `upstream/latest` is tracking `origin/upstream/latest` in my testing.
:::

Let's check out the `upstreamvcs/main` branch locally and call it `upstreamvcs-main`:

```{terminal}
:copy:
:host:
:dir: gnome-control-center
:user:
:input: git checkout -b upstreamvcs-main upstreamvcs/main

Branch 'upstreamvcs-main' set up to track remote branch 'main' from 'upstreamvcs'.
Switched to a new branch 'upstreamvcs-main'

> ```

Author: I guess it's trying to show the conventional way to check out the branch? That's how I'm explaining it for the time being.

Export the `../build-area/` directory before building with the `git-buildpackage` tool:

```{code-block} ini
:caption: `~/.gbp.conf`

[buildpackage]
export-dir = ../build-area/
```

For details, see the `gbp.conf(5)` and `gbp-buildpackage` man pages.

### Merging the changelog automatically

To merge the changelog automatically, enable the `dpkg-mergechangelogs` tool in the `~/.config/git/config` file:

```{code-block} ini
:caption: `~/.config/git/config`

[merge "dpkg-mergechangelogs"]  
    name = debian/changelog merge driver
    driver = dpkg-mergechangelogs -m %O %A %B %A
```

Create the `~/.config/git/attributes` file and set `dpkg-mergechangelogs` as the default strategy in it:

```{code-block} ini
:caption: `~/.config/git/attributes`

debian/changelog merge=dpkg-mergechangelogs
```

:::{warning}

> This automation might not work when Ubuntu packages use a higher epoch than the Debian ones. In that case, fix the changelog manually.

Reviewer: We don't have `epoch` described in [version string format](https://documentation.ubuntu.com/project/how-ubuntu-is-made/concepts/version-strings/) nor in the Glossary. In the meantime, let's link to Debian:

```suggestion
This automation might not work when Ubuntu packages use a higher [epoch](https://www.debian.org/doc/debian-policy/ch-controlfields.html#version) than the Debian ones. In that case, fix the changelog manually.
```
